<center><img src="car.jpg" alt="Parked car" width="400" height="300"></center>

Insurance companies invest a lot of [time and money](https://www.accenture.com/_acnmedia/pdf-84/accenture-machine-leaning-insurance.pdf) into optimizing their pricing and accurately estimating the likelihood that customers will make a claim. In many countries insurance it is a legal requirement to have car insurance in order to drive a vehicle on public roads, so the market is very large!

Knowing all of this, On the Road car insurance have requested your services in building a model to predict whether a customer will make a claim on their insurance during the policy period. As they have very little expertise and infrastructure for deploying and monitoring machine learning models, they've asked you to use simple Logistic Regression, identifying the single feature that results in the best performing model, as measured by accuracy.

They have supplied you with their customer data as a csv file called `car_insurance.csv`, along with a table detailing the column names and descriptions below.



## The dataset

| Column | Description |
|--------|-------------|
| `id` | Unique client identifier |
| `age` | Client's age: <br> <ul><li>`0`: 16-15</li><li>`1`: 26-39</li><li>`2`: 40-64</li><li>`3`: 65+</li></ul> |
| `gender` | Client's gender: <br> <ul><li>`0`: Female</li><li>`1`: Male</li></ul> |
| `driving_experience` | Years the client has been driving: <br> <ul><li>`0`: 0-9</li><li>`1`: 10-19</li><li>`2`: 20-29</li><li>`3`: 30+</li></ul> |
| `education` | Client's level of education: <br> <ul><li>`0`: No education</li><li>`1`: High school</li><li>`2`: University</li></ul> |
| `income` | Client's income level: <br> <ul><li>`0`: Poverty</li><li>`1`: Working class</li><li>`2`: Middle class</li><li>`3`: Upper class</li></ul> |
| `credit_score` | Client's credit score (between zero and one) |
| `vehicle_ownership` | Client's vehicle ownership status: <br><ul><li>`0`: Does not own their vehilce (paying off finance)</li><li>`1`: Owns their vehicle</li></ul> |
| `vehcile_year` | Year of vehicle registration: <br><ul><li>`0`: Before 2015</li><li>`1`: 2015 or later</li></ul> |
| `married` | Client's marital status: <br><ul><li>`0`: Not married</li><li>`1`: Married</li></ul> |
| `children` | Client's number of children |
| `postal_code` | Client's postal code | 
| `annual_mileage` | Number of miles driven by the client each year |
| `vehicle_type` | Type of car: <br> <ul><li>`0`: Sedan</li><li>`1`: Sports car</li></ul> |
| `speeding_violations` | Total number of speeding violations received by the client | 
| `duis` | Number of times the client has been caught driving under the influence of alcohol |
| `past_accidents` | Total number of previous accidents the client has been involved in |
| `outcome` | Whether the client made a claim on their car insurance (response variable): <br><ul><li>`0`: No claim</li><li>`1`: Made a claim</li></ul> |

In [1]:
# Import required modules
import pandas as pd
import numpy as np
from statsmodels.formula.api import logit

# Start coding!

# Reading the data

insurance = pd.read_csv('car_insurance.csv')
print(insurance)


          id  age  gender  ... duis past_accidents outcome
0     569520    3       0  ...    0              0     0.0
1     750365    0       1  ...    0              0     1.0
2     199901    0       0  ...    0              0     0.0
3     478866    0       1  ...    0              0     0.0
4     731664    1       1  ...    0              1     1.0
...      ...  ...     ...  ...  ...            ...     ...
9995  323164    1       0  ...    0              1     0.0
9996  910346    1       0  ...    0              0     0.0
9997  468409    1       1  ...    0              0     0.0
9998  903459    1       0  ...    0              1     1.0
9999  442696    1       0  ...    0              0     0.0

[10000 rows x 18 columns]


In [2]:
# Investigating the data for missing values

print(insurance.info())

print(insurance.isnull().sum())

# Checking the distribution of data in the columns of interest
print(insurance[['credit_score','annual_mileage']])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   10000 non-null  int64  
 1   age                  10000 non-null  int64  
 2   gender               10000 non-null  int64  
 3   driving_experience   10000 non-null  object 
 4   education            10000 non-null  object 
 5   income               10000 non-null  object 
 6   credit_score         9018 non-null   float64
 7   vehicle_ownership    10000 non-null  float64
 8   vehicle_year         10000 non-null  object 
 9   married              10000 non-null  float64
 10  children             10000 non-null  float64
 11  postal_code          10000 non-null  int64  
 12  annual_mileage       9043 non-null   float64
 13  vehicle_type         10000 non-null  object 
 14  speeding_violations  10000 non-null  int64  
 15  duis                 10000 non-null  

**Analysis after loading & investigating the data**

Upon loading the data we can see that there are missing values within the dataset while the distribution of the columns 'credit_score' & 'annual_mileage' while also having values that appear to be normally distributed.

To rectify i shall fill in the missing values and find they're means

In [3]:
# Replacing the missing values with means

insurance["credit_score"].fillna(insurance["credit_score"].mean(), inplace=True)
insurance["annual_mileage"].fillna(insurance["annual_mileage"].mean(), inplace=True)

print(insurance[['credit_score','annual_mileage']])

      credit_score  annual_mileage
0         0.629027    12000.000000
1         0.357757    16000.000000
2         0.493146    11000.000000
3         0.206013    11000.000000
4         0.388366    12000.000000
...            ...             ...
9995      0.582787    16000.000000
9996      0.522231    11697.003207
9997      0.470940    14000.000000
9998      0.364185    13000.000000
9999      0.435225    13000.000000

[10000 rows x 2 columns]


**Modelling**

Because we are trying to figure whether a claim will be made or not, a logistic regression algorithm is best suited to model the data and give us the best results needed to make the best business decisions. 

We now must prepare the data and pass it through the model to begin with.

In [4]:
# Prepairing the data for modelling
models = []
features = insurance.drop(columns=["id", "outcome"]).columns

# Loop through the features
for col in features:
    # Create a new Logistic Regression model
    model = logit(f"outcome ~ {col}", data=insurance).fit()

    # Save the model as a variable called models
    models.append(model)







Optimization terminated successfully.
         Current function value: 0.511794
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.615951
         Iterations 5
Optimization terminated successfully.
         Current function value: 0.467092
         Iterations 8
Optimization terminated successfully.
         Current function value: 0.603742
         Iterations 5
Optimization terminated successfully.
         Current function value: 0.531499
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.572557
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.552412
         Iterations 5
Optimization terminated successfully.
         Current function value: 0.572668
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.586659
         Iterations 5
Optimization terminated successfully.
         Current function value: 0.595431
  

**Accuracy calculation**

Now that the first phase of modelling has been completed we can now calculate the accuracy by using a confusion matrix upon the modelled data and then find the best feature and accuracy that tells us whether a claim will be made or not.

In [5]:
# performance measurement
accuracies = []

# Iterate over the index of models
for i in range(0, len(models)):
    # Compute the confusion matrix
    conf_matrix = models[i].pred_table()
    # True negatives
    tn = conf_matrix[0,0]
    # True positives
    tp = conf_matrix[1,1]
    # False negatives
    fn = conf_matrix[1,0]
    # False positives
    fp = conf_matrix[0,1]
    # Compute accuracy
    acc = (tn + tp) / (tn + fn + fp + tp)
    accuracies.append(acc)
    
# calculating accuracy and storing the models results
accuracy = (tn + tp) / (tn + tp + fn + fp)
accuracies.append(acc)

 




In [6]:
# Finding the best performing model
best_feature = features[accuracies.index(max(accuracies))]

# Create best_feature_df
best_feature_df = pd.DataFrame({"best_feature": best_feature,
                                "best_accuracy": max(accuracies)},
                                index=[0])
best_feature_df

best_feature  best_accuracy
0  driving_experience         0.7771

**Conclusion**

We have now calculated the accuracy of the model and detemined the best feature with the largest accurracy score.

We can see that the best feature is 'driving_experience' with an accuracy of 78%. This means that drivers with sufficient driving experience on record are more likely to make a claim on insurance than other determining factors.